File: python/diar_lab/plda.py
"""PLDA + x-vector transform (float64 LE; open DiariZen npz or legacy bin).

Matches BUT VBx recipe transform:
  x = l2_norm( lda.T @ l2_norm(x - mean1) - mean2 )
  fea = (x - plda_mu) @ plda_tr.T
"""

from __future__ import annotations

from dataclasses import dataclass
from pathlib import Path

import numpy as np


def _load_f64(path: Path) -> np.ndarray:
    data = path.read_bytes()
    assert len(data) % 8 == 0, path
    return np.frombuffer(data, dtype="<f8").copy()


def l2_norm(x: np.ndarray, axis: int = -1, eps: float = 1e-12) -> np.ndarray:
    n = np.linalg.norm(x, axis=axis, keepdims=True)
    return x / (n + eps)


@dataclass
class PldaTransform:
    mean1: np.ndarray  # (256,)
    lda: np.ndarray  # (256, 128)  — apply as (x-mean1) @ lda
    mean2: np.ndarray  # (128,)
    mu: np.ndarray  # (128,)
    tr: np.ndarray  # (128, 128)
    psi: np.ndarray  # (128,)

    @classmethod
    def load(cls, plda_dir: Path) -> "PldaTransform":
        """Load from DiariZen npz dir (plda.npz + xvec_transform.npz) or legacy bin dir."""
        plda_dir = Path(plda_dir)
        npz_plda = plda_dir / "plda.npz"
        npz_xvec = plda_dir / "xvec_transform.npz"
        if npz_plda.is_file() and npz_xvec.is_file():
            return cls.load_npz(npz_plda, npz_xvec)

        mean1 = _load_f64(plda_dir / "xvec_transform_mean1.bin")
        lda_flat = _load_f64(plda_dir / "xvec_transform_lda.bin")
        mean2 = _load_f64(plda_dir / "xvec_transform_mean2.bin")
        mu = _load_f64(plda_dir / "plda_mu.bin")
        tr_flat = _load_f64(plda_dir / "plda_tr_final.bin")
        psi = _load_f64(plda_dir / "plda_psi_final.bin")

        d_in, d_out = mean1.shape[0], mean2.shape[0]
        assert lda_flat.size == d_in * d_out
        lda = lda_flat.reshape(d_in, d_out)
        assert tr_flat.size == d_out * d_out
        tr = tr_flat.reshape(d_out, d_out)
        assert mu.shape == (d_out,) and psi.shape == (d_out,)
        # psi must be positive for VBx
        psi = np.maximum(psi, 1e-6)
        return cls(mean1=mean1, lda=lda, mean2=mean2, mu=mu, tr=tr, psi=psi)

    @classmethod
    def load_npz(cls, plda_npz: Path, xvec_npz: Path) -> "PldaTransform":
        plda = np.load(plda_npz)
        xvec = np.load(xvec_npz)
        mean1 = np.asarray(xvec["mean1"], dtype=np.float64).reshape(-1)
        mean2 = np.asarray(xvec["mean2"], dtype=np.float64).reshape(-1)
        lda = np.asarray(xvec["lda"], dtype=np.float64)
        if lda.shape == (mean2.shape[0], mean1.shape[0]):
            lda = lda.T  # (256, 128) expected
        mu = np.asarray(plda["mu"], dtype=np.float64).reshape(-1)
        tr = np.asarray(plda["tr"], dtype=np.float64)
        psi = np.maximum(np.asarray(plda["psi"], dtype=np.float64).reshape(-1), 1e-6)
        assert lda.shape == (mean1.shape[0], mean2.shape[0])
        assert tr.shape == (mean2.shape[0], mean2.shape[0])
        return cls(mean1=mean1, lda=lda, mean2=mean2, mu=mu, tr=tr, psi=psi)

    def xvec_transform(self, emb: np.ndarray) -> np.ndarray:
        """256-d embeddings → 128-d LDA space (BUT recipe)."""
        x = np.asarray(emb, dtype=np.float64)
        single = x.ndim == 1
        if single:
            x = x[None, :]
        x = l2_norm(x - self.mean1)
        x = x @ self.lda  # (N, 128)
        x = l2_norm(x - self.mean2)
        return x[0] if single else x

    def plda_space(self, x128: np.ndarray, lda_dim: int | None = None) -> np.ndarray:
        """128-d → PLDA projected features for VBx."""
        x = np.asarray(x128, dtype=np.float64)
        single = x.ndim == 1
        if single:
            x = x[None, :]
        fea = (x - self.mu) @ self.tr.T
        if lda_dim is not None:
            fea = fea[:, :lda_dim]
        return fea[0] if single else fea

    def transform_for_vbx(self, emb256: np.ndarray, lda_dim: int | None = None) -> np.ndarray:
        return self.plda_space(self.xvec_transform(emb256), lda_dim=lda_dim)

    def transform_for_ahc(self, emb256: np.ndarray) -> np.ndarray:
        """Features for cosine AHC (after xvec transform, before PLDA tr)."""
        return self.xvec_transform(emb256)

File: crates/diar-rs/src/plda.rs
//! PLDA + x-vector transform (float64 LE; open npz or legacy bin).
//!
//! Mirrors the Python lab `plda.py` / BUT VBx recipe:
//!   x = l2_norm( (l2_norm(x - mean1) @ lda) - mean2 )
//!   fea = (x - mu) @ tr.T

use std::fs;
use std::path::Path;

use byteorder::{ByteOrder, LittleEndian};

use crate::error::{Error, Result};

#[derive(Debug, Clone)]
pub struct Plda {
    pub mean1: Vec<f64>, // 256
    pub lda: Vec<f64>,   // 256 * 128 row-major (x @ lda)
    pub mean2: Vec<f64>, // 128
    pub mu: Vec<f64>,    // 128
    pub tr: Vec<f64>,    // 128 * 128 row-major
    pub psi: Vec<f64>,   // 128
    pub d_in: usize,
    pub d_out: usize,
}

fn load_f64_le(path: &Path) -> Result<Vec<f64>> {
    let bytes = fs::read(path).map_err(|e| {
        if e.kind() == std::io::ErrorKind::NotFound {
            Error::MissingModel(path.to_path_buf())
        } else {
            Error::Io(e)
        }
    })?;
    if bytes.len() % 8 != 0 {
        return Err(Error::Plda(format!(
            "{}: size {} not multiple of 8",
            path.display(),
            bytes.len()
        )));
    }
    let n = bytes.len() / 8;
    let mut out = vec![0.0f64; n];
    LittleEndian::read_f64_into(&bytes, &mut out);
    Ok(out)
}

fn l2_norm_row(x: &mut [f64]) {
    let mut s = 0.0;
    for v in x.iter() {
        s += *v * *v;
    }
    let n = s.sqrt().max(1e-12);
    for v in x.iter_mut() {
        *v /= n;
    }
}

impl Plda {
    pub fn load(dir: impl AsRef<Path>) -> Result<Self> {
        let dir = dir.as_ref();
        let mean1 = load_f64_le(&dir.join("xvec_transform_mean1.bin"))?;
        let lda = load_f64_le(&dir.join("xvec_transform_lda.bin"))?;
        let mean2 = load_f64_le(&dir.join("xvec_transform_mean2.bin"))?;
        let mu = load_f64_le(&dir.join("plda_mu.bin"))?;
        let tr = load_f64_le(&dir.join("plda_tr_final.bin"))?;
        let mut psi = load_f64_le(&dir.join("plda_psi_final.bin"))?;

        let d_in = mean1.len();
        let d_out = mean2.len();
        if lda.len() != d_in * d_out {
            return Err(Error::Plda(format!(
                "lda len {} != {}x{}",
                lda.len(),
                d_in,
                d_out
            )));
        }
        if tr.len() != d_out * d_out || mu.len() != d_out || psi.len() != d_out {
            return Err(Error::Plda("mu/tr/psi shape mismatch".into()));
        }
        for p in psi.iter_mut() {
            *p = p.max(1e-6);
        }
        Ok(Self {
            mean1,
            lda,
            mean2,
            mu,
            tr,
            psi,
            d_in,
            d_out,
        })
    }

    /// Single 256-d embedding → 128-d AHC space.
    pub fn xvec_transform(&self, emb: &[f64]) -> Result<Vec<f64>> {
        if emb.len() != self.d_in {
            return Err(Error::Plda(format!(
                "emb dim {} != {}",
                emb.len(),
                self.d_in
            )));
        }
        let mut x: Vec<f64> = emb
            .iter()
            .zip(self.mean1.iter())
            .map(|(a, b)| a - b)
            .collect();
        l2_norm_row(&mut x);
        // x @ lda  → 128
        let mut y = vec![0.0f64; self.d_out];
        for j in 0..self.d_out {
            let mut s = 0.0;
            for i in 0..self.d_in {
                s += x[i] * self.lda[i * self.d_out + j];
            }
            y[j] = s;
        }
        for (v, m) in y.iter_mut().zip(self.mean2.iter()) {
            *v -= m;
        }
        l2_norm_row(&mut y);
        Ok(y)
    }

    /// 128-d → PLDA projected features for VBx: (x - mu) @ tr.T
    pub fn plda_space(&self, x128: &[f64]) -> Result<Vec<f64>> {
        if x128.len() != self.d_out {
            return Err(Error::Plda(format!(
                "x128 dim {} != {}",
                x128.len(),
                self.d_out
            )));
        }
        let mut centered = vec![0.0f64; self.d_out];
        for i in 0..self.d_out {
            centered[i] = x128[i] - self.mu[i];
        }
        // fea_j = sum_i centered_i * tr[j, i]  == (x-mu) @ tr.T
        let mut fea = vec![0.0f64; self.d_out];
        for j in 0..self.d_out {
            let mut s = 0.0;
            for i in 0..self.d_out {
                s += centered[i] * self.tr[j * self.d_out + i];
            }
            fea[j] = s;
        }
        Ok(fea)
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use std::path::PathBuf;

    /// Open PLDA (.bin converted from the DiariZen npz) under models/plda.
    fn open_plda() -> Option<PathBuf> {
        let p = PathBuf::from(env!("CARGO_MANIFEST_DIR")).join("../../models/plda");
        if p.join("plda_mu.bin").is_file() {
            Some(p)
        } else {
            None
        }
    }

    #[test]
    fn load_open_plda_shapes() {
        let Some(dir) = open_plda() else {
            eprintln!("skip: open plda not under models/plda");
            return;
        };
        let p = Plda::load(&dir).expect("load");
        assert_eq!(p.d_in, 256);
        assert_eq!(p.d_out, 128);
        assert_eq!(p.psi.len(), 128);
        assert!(p.psi.iter().all(|&v| v > 0.0));
    }

    #[test]
    fn xvec_roundtrip_dims() {
        let Some(dir) = open_plda() else {
            return;
        };
        let p = Plda::load(&dir).unwrap();
        let emb = vec![0.01f64; 256];
        let x = p.xvec_transform(&emb).unwrap();
        assert_eq!(x.len(), 128);
        let f = p.plda_space(&x).unwrap();
        assert_eq!(f.len(), 128);
        let norm: f64 = x.iter().map(|v| v * v).sum::<f64>().sqrt();
        assert!((norm - 1.0).abs() < 1e-6, "xvec should be l2-normalized");
    }
}
